File: C++ Codes/to_do_digit_dp.cpp
#include<bits/stdc++.h>

using namespace std;

#define int long long int
#define ld long double
#define F first
#define S second
#define pb push_back

const int N = 100005;

string s;
int dp[2001][2][2000], m, mod = 1e9 + 7;

int work(int pos, bool last, int modpro) {
	if (pos == s.size()) {
		if (modpro == 0) {
			return 1;
		}
		return 0;
	}
	if (dp[pos][last][modpro] != -1) {
		return dp[pos][last][modpro];
	}
	int till = (last ? s[pos] - '0' : 9);
	int ans = 0;
	for (int i = 0; i <= till; i++) {
		ans += work(pos + 1, last && (i == s[pos] - '0'), (modpro * 10 + i) % m);
		ans %= mod;
	}
	return dp[pos][last][modpro] = ans % mod;
}

int solve(int n) {
	s = to_string(n);
	memset(dp, -1, sizeof(dp));
	int ans = work(0, 1, 0);
	return ans;
}

int32_t main()
{
	ios_base:: sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);
	freopen("debug.txt", "w", stderr);
#endif
	int t; cin >> t; while (t--)
	{
		int i, j, k, n, ans = 0, cnt = 0, sum = 0;
		cin >> i >> j >> m;
		cout << solve(j) - solve(i - 1);
	}
}

File: Java Codes/ToDoDigitDp.java
package DP;

import java.util.Arrays;
import java.util.Locale;
import java.util.Scanner;

public class ToDoDigitDp {



    static final int N = 100005;

    static String s;
    static int[][][] dp = new int[2001][2][2000];
    static int m, mod = (int) (1e9 + 7);

    static int work(int pos, boolean last, int modpro) {
        if (pos == s.length()) {
            if (modpro == 0) {
                return 1;
            }
            return 0;
        }
        if (dp[pos][last ? 1 : 0][modpro] != -1) {
            return dp[pos][last? 1 : 0][modpro];
        }
        int till = (last ? s.charAt(pos) - '0' : 9);
        int ans = 0;
        for (int i = 0; i <= till; i++) {
            ans += work(pos + 1, last && (i == s.charAt(pos) - '0'), (modpro * 10 + i) % m);
            ans %= mod;
        }
        return dp[pos][last ? 1: 0][modpro] = ans % mod;
    }

    static int solve(int n) {
        s = String.valueOf(n);

        for(int i = 0; i < dp.length; i++){
            for(int j = 0; j < dp[0].length; j++)
                Arrays.fill(dp[i][j], -1);
        }
        int ans = work(0, true, 0);
        return ans;
    }

    public static void main(String[] args) {

        Scanner scn = new Scanner(System.in);
        int t = scn.nextInt();
        while (t-- > 0)
        {
            int i = scn.nextInt(), j=scn.nextInt();
            m=scn.nextInt();
            System.out.println(solve(j) - solve(i - 1));
        }
    }
}
